README.md



sp-json-logger is **a simple and fast JSON logging library** for node.js services. It is a wrapper module which adds more functionality and conventions to the bunyan logger.
Readme for [version 1.0.9 ](https://github.com/ShoppinPal/sp-json-logger/blob/develop/Readme-1.0.9.md)
# Installation

```sh
npm install sp-json-logger
```

```js
var log = require('sp-json-logger')(); // without passing config object
var log = require('sp-json-logger')({fileName: __filename}); // passing config object
log.info('hi');
```
**Important:** It is important to have **NAME**, **APPLICATION**, **PROGRAM** and **LANGUAGE** env variables set up before requiring the module!

# Conventions followed while logging:

You need to set the following environment variables for logger to work without our logger module throwing errors ( We can change this behavior if we want, but it will help us create a habit to specify env variables which can be used to identify our application in multi app environment where we have multiple modules split up into different microservices each having its own language and platform)

- **NAME:** 'Name of the logger' else will default to 'sp-json-logger'
- **APPLICATION:** 'Name of applicaton'
- **PROGRAM:** 'Name of module' // eg: scheduler-worker or scheduler-api
- **LANGUAGE:** 'Programming language used' // eg: javascript/php/go
- **NODE_ENV:** 'environment' // note: Pretty print is supported on local environment setting, if using staging or production, you won’t get pretty json output to console.
- **SP_DISABLE_LOGS:** Use this env variable as `SP_DISABLE_LOGS=true` to disable all logs from `sp-json-logger`.
- ***SP_PRETTY_PRINT:*** Set its value as `SP_PRETTY_PRINT=true` to enable pretty print of json.
If logging only string message, use the following format:

`logger.debug({message: 'Your string here...'});` or `logger.debug('Your string here...');`

Also use `logger.tag('CONNECTION').debug({ message: 'Successfully connected' });`
If logging a JSON Object, use following format:
```js
logger.tag('AUDIT').debug({
	type: 'TYPE of LOG',  // eg: AUDIT Created
	objectName: objectValue,
});
```

# Different log levels: (Log levels corresponds to different integer values)
Source: [https://www.npmjs.com/package/bunyan#levels](https://www.npmjs.com/package/bunyan#levels)

- "fatal" (60): The service/app is going to stop or become unusable now. An operator should definitely look into this soon.
- "error" (50): Fatal for a particular request, but the service/app continues servicing other requests. An operator should look at this soon(ish).
- "warn" (40): A note on something that should probably be looked at by an operator eventually.
- "info" (30): Detail on regular operation.
- "debug" (20): Anything else, i.e. too verbose to be included in "info" level.
- "trace" (10): Logging from external libraries used by your app or very detailed application logging.


# Key Notes:

- Specify name of the logger in the NAME env variable to be meaningful.
- Use tags wherever applicable i.e `logger.tag(‘TAG’).error(err);`
- Order version required you to log by explcitly stating parent object name i.e **log** as per following example:


```js
// Older version (1.0.9)
logger.debug({log: {message: 'hello world'}});
// Newer version
logger.debug('hello world');
logger.debug({message: 'hello world'});
logger.error('hello world');
logger.error({message: 'hello world'});
```
- You can set parent object name by utilizing the method `setParentObjectName('String')` explained below.
- As far as possible use `logger.tag('TAG');` field to describe the event for which we are logging.
- If you just need to log a string, you can create it as follows: 

	`logger.debug('your string');`
- To enable pretty print, use the environment variable `SP_PRETTY_PRINT=true`.

# Local environment vs Staging/Production environment:

It was observed that printing json through bunyan was not readable for developers. In short it did not have pretty print functionality built in. 
We solved this problem by using the following module: https://github.com/WoLfulus/bunyans
Above module provided a new output stream that pretty prints json object to console.

So during local environment, we can enjoy pretty print. During staging and production, we get normal json output which helps us to reduce the impact of performance for pretty printing.

Important: Pretty print stream is a huge performance overhead, so it is recommended to use it only for local/development environment (if you ever want to modify bunyanLogger.js file inside sp-json-logger module)
# Testing:

* To test without publishing, go to the `PROJECT_ROOT` directory and run `npm link`. Sample output:

	```
	$ npm link
	~/.nvm/versions/v7.10.0/lib/node_modules/sp-json-logger -> ~/path/to/<PROJECT_ROOT>
	```
* First start logstash and elasticsearch instances
	* Run the command from root directory: `docker-compose up`
	* Once logstash, elasticsearch and kibana instances are up and running, build the docker image for tests: 
	
		`docker build ./ --tag tests`
	* Figure out the default network running the ELK stack:

		```
		$ docker network ls
		NETWORK ID          NAME                    DRIVER              SCOPE
		009031fa9b3f        spjsonlogger_default    bridge              local
		```
	* Attach and run the tests on that network: 
	
		`docker run --network=spjsonlogger_default tests`
	* View test results on kibana: `http://localhost:5601`
        * You must run the tests first before coming here if you want to see meaningful data
		* `Time Filter field name` should be set to `time`
		* Jump to the `Discover` tab on top left and you should see the logs generated by the tests run.

* (Another way) Now you can create and run tests:
	* `node test/test.local.1.js`

		```
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":30,"application":"","program":"","language":"","log":{"messag
		e":"hi"},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.923Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","log":{"message":"Your string here..."},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.928Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","log":{"message":"Successfully connected"},"tag":"myTagA","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.931Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","log":{"type":
		"AUDIT","habitable":{"planets":["mars","earth"]}},"tag":"myTagB","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","
		msg":"","time":"2018-03-28T18:22:45.932Z","v":0}

		Regex with JSON.stingify(object, replacer)

		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","log":{"query"
		:"{\"sku\":\"/^BA1262$/i\"}"},"tag":"Regex","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.942Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","log":{"query":"{\"query\":[{\"sku\":\"/^BA1262$/i\"},{\"sku\":\"/^BRAT$/i\"}]}"},"tag":"RegExArray","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.942Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","log":{"sku":"/^BA1262$/"},"tag":"parse(boolean)","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.944Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","log":{"query":[{"sku":"/^BA1262$/"},{"sku":"/^BRAT$/"}]},"tag":"RegExArray parse(boolean)","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.945Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":50,"application":"","program":"","language":"","err":{"message":"the earth is flat","name":"Error","stack":"Error: the earth is flat\n    at Object.<anonymous> (/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js:34:13)\n    at Module._compile (module.js:570:32)\n    at Object.Module._extensions..js (module.js:579:10)\n    at Module.load (module.js:487:32)\n    at tryModuleLoad (module.js:446:12)\n    at Function.Module._load (module.js:438:3)\n    at Module.runMain (module.js:604:10)\n    at run (bootstrap_node.js:389:7)\n    at startup (bootstrap_node.js:149:9)\n    at bootstrap_node.js:504:3"},"log":{},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"the earth is flat","time":"2018-03-28T18:22:45.946Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":50,"application":"","program":"","language":"","err":{"message":"the earth is round :p","name":"discovery","stack":"Some stack here....."},"log":{"message":"the earth is round :p","name":"discovery"},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"the earth is round :p","time":"2018-03-28T18:22:45.947Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":30,"application":"","program":"","language":"","dump":{"message":"Hi...."},"tag":"ParentObject name changed","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.947Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","dump":{"arg":"some arg"},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.948Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","dump":{"demo":{"text":"hello world"},"someObj":{"p":"content.."}},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.948Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":20,"application":"","program":"","language":"","dump":{"demo":{"number":0,"obj":{"a":"a","b":"b"}}},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-03-28T18:22:45.949Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":50,"application":"","program":"","language":"","err":{"message":"the earth is round :p","name":"discovery","stack":"Some stack here....."},"dump":{"message":"Some error"},"tag":"hello error","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"the earth is round :p","time":"2018-03-28T18:22:45.949Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":29062,"level":50,"application":"","program":"","language":"","err":{"message":"the earth is round :p","name":"discovery","stack":"Some stack here....."},"log":{"message":"My custom error message","functionName":"hello()"
		},"fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"the earth is round :p","time":"2018-03-28T18:22:45.949Z","v":0}
		{"name":"sp-json-logger","hostname":"3076905ec882","pid":5,"level":50,"application":"tests","program":"sp-json-logger","language":"javascript","log":{"message":"My custom error message","functionName":"error()"},"tag":"Not-Error-Instance","fileName":"/app/test/test.1.js","msg":"","time":"2018-03-29T09:54:30.393Z","v":0}
		{"name":"sp-json-logger","hostname":"Yogeshs-MacBook-Air.local","pid":95722,"level":50,"application":"","program":"","language":"","log":{"message":"hello error"},"tag":"Not-Error-Instance","fileName":"/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js","msg":"","time":"2018-04-29T11:53:51.606Z","v":0}
		```
	* `SP_PRETTY_PRINT=true node test/test.local.1.js`

		```
		[2018-03-28T18:34:43.503Z]  INFO: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="")

		--
		log: {
		"message": "hi"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.511Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="")

		--
		log: {
		"message": "Your string here..."
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.513Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag=myTagA)

		--
		log: {
		"message": "Successfully connected"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.514Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag=myTagB)

		--
		log: {
		"type": "AUDIT",
		"habitable": {
			"planets": [
			"mars",
			"earth"
			]
		}
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js

		Regex with JSON.stingify(object, replacer)

		[2018-03-28T18:34:43.522Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag=Regex)

		--
		log: {
		"query": "{\"sku\":\"/^BA1262$/i\"}"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.523Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag=RegExArray)

		--
		log: {
		"query": "{\"query\":[{\"sku\":\"/^BA1262$/i\"},{\"sku\":\"/^BRAT$/i\"}]}"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.525Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag=parse(boolean))

		--
		log: {
		"sku": "/^BA1262$/"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.525Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag="RegExArray parse(boolean)")

		--
		log: {
		"query": [
			{
			"sku": "/^BA1262$/"
			},
			{
			"sku": "/^BRAT$/"
			}
		]
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.528Z] ERROR: sp-json-logger/29239 on Yogeshs-MacBook-Air.local: the earth is flat (application="", program="", language="",log={})
		Error: the earth is flat
			at Object.<anonymous> (/Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js:34:13)
			at Module._compile (module.js:570:32)
			at Object.Module._extensions..js (module.js:579:10)
			at Module.load (module.js:487:32)
			at tryModuleLoad (module.js:446:12)
			at Function.Module._load (module.js:438:3)
			at Module.runMain (module.js:604:10)
			at run (bootstrap_node.js:389:7)
			at startup (bootstrap_node.js:149:9)
			at bootstrap_node.js:504:3
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.529Z] ERROR: sp-json-logger/29239 on Yogeshs-MacBook-Air.local: the earth is round :p (application="", program="", language="")
		Some stack here.....
		--
		log: {
		"message": "the earth is round :p",
		"name": "discovery"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.530Z]  INFO: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag="ParentObject name changed")

		--
		dump: {
		"message": "Hi...."
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.530Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="")

		--
		dump: {
		"arg": "some arg"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.531Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="")

		--
		dump: {
		"demo": {
			"text": "hello world"
		},
		"someObj": {
			"p": "content.."
		}
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.532Z] DEBUG: sp-json-logger/29239 on Yogeshs-MacBook-Air.local:  (application="", program="", language="")

		--
		dump: {
		"demo": {
			"number": 0,
			"obj": {
			"a": "a",
			"b": "b"
			}
		}
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.539Z] ERROR: sp-json-logger/29239 on Yogeshs-MacBook-Air.local: the earth is round :p (application="", program="", language="", tag="hello error")
		Some stack here.....
		--
		dump: {
		"message": "Some error"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-28T18:34:43.540Z] ERROR: sp-json-logger/29239 on Yogeshs-MacBook-Air.local: the earth is round :p (application="", program="", language="")
		Some stack here.....
		--
		log: {
		"message": "My custom error message",
		"functionName": "hello()"
		}
		--
		fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-03-29T09:58:27.409Z] ERROR: sp-json-logger/35398 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag=Not-Error-Instance)
    	--
    	log: {
      	"message": "My custom error message",
      	"functionName": "hello()"
    	}
    	--
    	fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		[2018-04-29T11:53:00.841Z] ERROR: sp-json-logger/95687 on Yogeshs-MacBook-Air.local:  (application="", program="", language="", tag=Not-Error-Instance)
    	--
    	log: {
      	"message": "hello error"
    	}
    	--
    	fileName: /Users/yogeshjadhav/Documents/sp-json-logger/test/test.local.1.js
		```
# setParentObjectName('String')

If we are not setting the parent object name, it defaults to `log`. But if we use `setParentObjectName('dump')` then the parent object name will become `dump`.

Eg: Without `setParentObjectName('String')`

```
logger.info('hello world');
```
It will output log as : 

`{..., {log: {message: 'hello world'},... }`

Eg: With `setParentObjectName('String')`

```
logger.setParentObjectName('dump');
logger.info('hello world with dump');

# Again calling setParentObjectName('String')
logger.setParentObjectName('payload');
logger.info('hello world with payload');
```
It will output log as : 
```
{..., {dump: {message: 'hello world with dump'},... }
{..., {payload: {message: 'hello world with payload'},... }
```
# Filtering objects while logging

To filter objects, you can use `filter({object: ['key1', 'key2']})` function.

Syntax: 

	`logger.filter({someObject: ['key1', 'key2']}).debug({someObject: object});` 

Here `someObject` is the object whose key needs to be filtered. The value for `someObject` is the array of keys which we want to keep in it during logging.


changelog.md
v2.1.3
- This release brings a set of new minor feature patches as given below
    - You can log an error string using the syntax: `logger.error('some error');`
    - If you want to enable pretty print, then use the new env variable: `SP_PRETTY_PRINT=true` instead of earlier `NODE_ENV=local`.
    - If you want to disable all the logs from `sp-json-logger`, then use the env variable `SP_DISABLE_LOGS=true`.

v2.1.2
- Intelligent error object detection. If we pass custom object inside error log, it will be encapsulated into `log` object or whichever object name is set using `setParentObjectName()`.
But if we are passing an error object instance, it will be encapsulated inside `err` object.

v2.1.1
- Error logging output changed. Previously any property logged other than `err` was at the same level as the `err` object.
Now it has been changed to be wrapped inside a parentObject (log by default).
    
v2.1.0
- Pass config object while instantiating logger object.
- Currently config object only suports `fileName` property.

v2.0.1
- Fix for error object not printing stack trace

v2.0


- New syntax for making life of developer easier.
- Now do `logger.debug('hello world)` for single string logging. Replaces older syntax `logger.debug({log: { message: 'hello world' }});`.
- For error logging, do `logger.error(error)`. Older syntax: `logger.error({err: error});` is deprecated.
- New function `parse(boolean)` added to log regex. It will scan for regex object in your JSON log and stringify it so that it can be logged properly. This operation is heavy on cpu so use alternative method like `replacer function in JSON.stringify()` to convert `regex object` to `string`.
- New function `setParentObjectName('string')` added to set parent log object name. By default it is set to `log`.
- New function `filter({ objectToFilter: ['key1', 'key2'] });` added. You can now filter (cherry pick) properties of an object to log.

src/prettystream.js
"use strict";

var Stream = require('stream').Stream;
var util = require('util');
var format = util.format;
var http = require('http');

var colors = {
    'bold': [1, 22],
    'italic': [3, 23],
    'underline': [4, 24],
    'inverse': [7, 27],
    'white': [37, 39],
    'grey': [90, 39],
    'black': [30, 39],
    'blue': [34, 39],
    'cyan': [36, 39],
    'green': [32, 39],
    'magenta': [35, 39],
    'red': [31, 39],
    'yellow': [33, 39]
};

var defaultOptions = {
    mode: 'long', //short, long, dev, none
    useColor: 'auto' // true, false, 'auto' (true if output is a TTY)
};

var levelFromName = {
    'trace': 10,
    'debug': 20,
    'info': 30,
    'warn': 40,
    'error': 50,
    'fatal': 60
};

var colorFromLevel = {
    10: 'grey',     // TRACE
    20: 'grey',     // DEBUG
    30: 'cyan',     // INFO
    40: 'magenta',  // WARN
    50: 'red',      // ERROR
    60: 'inverse'  // FATAL
};

var nameFromLevel = {};
var upperNameFromLevel = {};
var upperPaddedNameFromLevel = {};
Object.keys(levelFromName).forEach(function (name) {
    var lvl = levelFromName[name];
    nameFromLevel[lvl] = name;
    upperNameFromLevel[lvl] = name.toUpperCase();
    upperPaddedNameFromLevel[lvl] = (name.length === 4 ? ' ' : '') + name.toUpperCase();
});

//---- Internal support stuff

// http://stackoverflow.com/questions/4816099/chrome-sendrequest-error-typeerror-converting-circular-structure-to-json
var censor = function (censor) {
    return (function () {
        var keyList, valueOrderedSet;
        keyList = [];
        valueOrderedSet = [];
        return function (key, value) {
            var index;
            index = valueOrderedSet.indexOf(value);
            // if value is object and not found in valueOrderedSet
            if (typeof value === 'object' && index > -1) {
                // Note: null can also be circular!
                if (value === null) {
                    return 'null';
                }
                return '[Circular] ' + keyList[index];
            }
            keyList.push(key);
            valueOrderedSet.push(value);
            return value;
        }
    })(censor);
};

var JSON_stringify = function (/*obj, fn, format*/) {
    var args = [].slice.call(arguments)

    args[1] = censor(args[0])
    return JSON.stringify.apply(null, args);
};

function PrettyStream(opts) {
    var self = this, options = {};

    if (opts) {
        Object.keys(opts).forEach(function (key) {
            options[key] = {
                value: opts[key],
                enumerable: true,
                writable: true,
                configurable: true
            };
        });
    }

    var config = Object.create(defaultOptions, options);

    this.readable = true;
    this.writable = true;
    this.isTTY = false;

    Stream.call(this);

    function stylize(str, color) {
        if (!str) {
            return '';
        }

        if (config.useColor === false ||
            (config.useColor === 'auto' && !self.isTTY)) {
            return str;
        }

        if (!color) {
            color = 'white';
        }

        var codes = colors[color];
        if (codes) {
            return '\x1B[' + codes[0] + 'm' + str +
                '\x1B[' + codes[1] + 'm';
        }
        return str;
    }

    function indent(s) {
        return '    ' + s.split(/\r?\n/).join('\n    ');
    }

    function extractTime(rec) {
        var time = (typeof rec.time === 'object') ? rec.time.toISOString() : rec.time;

        if ((config.mode === 'short' || config.mode === 'dev') && time[10] == 'T') {
            return stylize(time.substr(11));
        }
        return stylize(time);
    }

    function extractName(rec) {
        var name = rec.name;

        if (rec.component) {
            name += '/' + rec.component;
        }

        if (config.mode !== 'short' && config.mode !== 'dev') {
            name += '/' + rec.pid;
        }

        return name;
    }

    function extractLevel(rec) {
        var level = (upperPaddedNameFromLevel[rec.level] || 'LVL' + rec.level);
        return stylize(level, colorFromLevel[rec.level]);
    }

    function extractSrc(rec) {
        var src = '';
        if (rec.src && rec.src.file) {
            if (rec.src.func) {
                src = format('(%s:%d in %s)', rec.src.file, rec.src.line, rec.src.func);
            } else {
                src = format('(%s:%d)', rec.src.file, rec.src.line);
            }
        }
        return stylize(src, 'green');
    }

    function extractHost(rec) {
        return rec.hostname || '<no-hostname>';
    }

    function isSingleLineMsg(rec) {
        return rec.msg.indexOf('\n') === -1;
    }

    function extractMsg(rec) {
        return stylize(rec.msg, 'cyan');
    }

    function extractReqDetail(rec) {
        if (rec.req && typeof (rec.req) === 'object') {
            var req = rec.req;
            var headers = req.headers;

            var str = format('%s %s HTTP/%s%s%s',
                req.method,
                req.url,
                req.httpVersion || '1.1',
                (req.remoteAddress ? "\nremote: " + req.remoteAddress + ":" + req.remotePort : ""),
                (headers ?
                    '\n' + Object.keys(headers).map(function (h) {
                        return h + ': ' + headers[h];
                    }).join('\n') :
                    '')
            );

            if (req.body) {
                str += '\n\n' + (typeof (req.body) === 'object' ? JSON_stringify(req.body, null, 2) : req.body);
            }
            if (req.trailers && Object.keys(req.trailers) > 0) {
                str += '\n' + Object.keys(req.trailers).map(function (t) {
                    return t + ': ' + req.trailers[t];
                }).join('\n');
            }

            var skip = ['headers', 'url', 'httpVersion', 'body', 'trailers', 'method', 'remoteAddress', 'remotePort'];

            var extras = {};

            Object.keys(req).forEach(function (k) {
                if (skip.indexOf(k) === -1) {
                    extras['req.' + k] = req[k];
                }
            });

            return {
                details: [str],
                extras: extras
            };
        }
    }

    function genericRes(res) {
        var s = '';

        if (res.statusCode) {
            s += format('HTTP/1.1 %s %s\n', res.statusCode, http.STATUS_CODES[res.statusCode]);
        }

        // Handle `res.header` or `res.headers` as either a string or
        // and object of header key/value pairs. Prefer `res.header` if set
        // (TODO: Why? I don't recall. Typical of restify serializer?
        // Typical JSON.stringify of a core node HttpResponse?)
        var headers;
        if (res.header !== undefined) {
            headers = res.header;
        } else if (res.headers !== undefined) {
            headers = res.headers;
        }

        if (!headers) {
            // pass through
        } else if (typeof (headers) === 'string') {
            s += headers.trimRight();
        } else {
            s += Object.keys(headers).map(
                function (h) { return h + ': ' + headers[h]; }).join('\n');
        }

        if (res.body) {
            s += '\n\n' + (typeof (res.body) === 'object' ? JSON_stringify(res.body, null, 2) : res.body);
        }
        if (res.trailer) {
            s += '\n' + res.trailer;
        }

        var skip = ['header', 'statusCode', 'headers', 'body', 'trailer'];

        var extras = {};

        Object.keys(res).forEach(function (k) {
            if (skip.indexOf(k) === -1) {
                extras['res.' + k] = res[k];
            }
        });

        return {
            details: [s],
            extras: extras
        };
    }

    function extractResDetail(rec) {
        if (rec.res && typeof (rec.res) === 'object') {
            return genericRes(rec.res);
        }
    }

    function extractClientReqDetail(rec) {
        if (rec.client_req && typeof (rec.client_req) === 'object') {
            var client_req = rec.client_req;

            var headers = client_req.headers;
            var hostHeaderLine = '';
            var s = '';

            if (client_req.address) {
                hostHeaderLine = 'Host: ' + client_req.address;

                if (client_req.port) {
                    hostHeaderLine += ':' + client_req.port;
                }

                hostHeaderLine += '\n';
            }

            s += format('%s %s HTTP/%s\n%s%s', client_req.method,
                client_req.url,
                client_req.httpVersion || '1.1',
                hostHeaderLine,
                (headers ?
                    Object.keys(headers).map(
                        function (h) {
                            return h + ': ' + headers[h];
                        }).join('\n') :
                    ''));

            if (client_req.body) {
                s += '\n\n' + (typeof (client_req.body) === 'object' ?
                    JSON_stringify(client_req.body, null, 2) :
                    client_req.body);
            }

            var skip = ['headers', 'url', 'httpVersion', 'body', 'trailers', 'method', 'remoteAddress', 'remotePort'];

            var extras = {};

            Object.keys(client_req).forEach(function (k) {
                if (skip.indexOf(k) === -1) {
                    extras['client_req.' + k] = client_req[k];
                }
            });

            return {
                details: [s],
                extras: extras
            };
        }
    }

    function extractClientResDetail(rec) {
        if (rec.client_res && typeof (rec.client_res) === 'object') {
            return genericRes(rec.client_res);
        }
    }

    function extractError(rec) {
        if (rec.err && rec.err.stack) {
            return rec.err.stack;
        }
    }

    function extractCustomDetails(rec) {
        var skip = ['name', 'hostname', 'pid', 'level', 'component', 'msg', 'time', 'v', 'src', 'err', 'client_req', 'client_res', 'req', 'res'];

        var details = [];
        var extras = {};

        Object.keys(rec).forEach(function (key) {
            if (skip.indexOf(key) === -1) {
                var value = rec[key];
                if (typeof value === 'undefined') value = '';
                var stringified = false;
                if (typeof value === 'function') {
                    value = '[Function]';
                    stringified = true;
                } else if (typeof value !== 'string') {
                    value = JSON_stringify(value, null, 2);
                    stringified = true;
                }
                if (value.indexOf('\n') !== -1 || value.length > 50) {
                    details.push(key + ': ' + value);
                } else if (!stringified && (value.indexOf(' ') != -1 || value.length === 0)) {
                    extras[key] = JSON_stringify(value);
                } else {
                    extras[key] = value;
                }
            }
        });

        return {
            details: details,
            extras: extras
        };
    }

    function applyDetails(results, details, extras) {
        if (results) {
            results.details.forEach(function (d) {
                details.push(indent(d));
            });
            Object.keys(results.extras).forEach(function (k) {
                extras.push(k + '=' + results.extras[k]);
            });
        }
    }

    this.formatRecord = function formatRecord(rec) {
        var details = [];
        var extras = [];

        var time = extractTime(rec);
        var level = extractLevel(rec);
        var name = extractName(rec);
        var host = extractHost(rec);
        var src = extractSrc(rec);

        var msg = isSingleLineMsg(rec) ? extractMsg(rec) : '';
        if (!msg) {
            details.push(indent(extractMsg(rec)));
        }

        var error = extractError(rec);
        if (error) {
            details.push(indent(error));
        }

        if (rec.req) { applyDetails(extractReqDetail(rec), details, extras); }
        if (rec.res) { applyDetails(extractResDetail(rec), details, extras); }
        if (rec.client_req) { applyDetails(extractClientReqDetail(rec), details, extras); }
        if (rec.client_res) { applyDetails(extractClientResDetail(rec), details, extras); }

        applyDetails(extractCustomDetails(rec), details, extras);

        extras = stylize(
            (extras.length ? ' (' + extras.join(', ') + ')' : ''), 'grey');
        details = stylize(
            (details.length ? details.join('\n    --\n') + '\n' : ''), 'grey');

        if (typeof config.mode === 'function') {
            return config.mode(time, level, name, host, src, msg, extras, details);
        }

        if (config.mode === 'short') {
            return format('[%s] %s %s: %s%s\n%s',
                time,
                level,
                name,
                msg,
                extras,
                details);
        }
        else if (config.mode === 'dev') {
            return format('%s %s %s %s: %s%s\n%s',
                time,
                level,
                name,
                src,
                msg,
                extras,
                details);
        }
        else if (config.mode === 'none') {
            return format('%s%s\n%s',
                msg,
                extras,
                details);
        }
        else { //if (config.mode === 'long'){
            return format('[%s] %s: %s on %s%s: %s%s\n%s',
                time,
                level,
                name,
                host,
                src,
                msg,
                extras,
                details);
        }
    };
}

util.inherits(PrettyStream, Stream);

PrettyStream.prototype.write = function write(data) {
    if (typeof data === 'string') {
        this.emit('data', this.formatRecord(JSON.parse(data)));
    } else if (typeof data === 'object') {
        this.emit('data', this.formatRecord(data));
    }
    return true;
};

PrettyStream.prototype.end = function end() {
    this.emit('end');
    return true;
};

// Track if we're piping into a TTY.
PrettyStream.prototype.pipe = function (dest, options) {
    this.isTTY = dest.isTTY;
    return Stream.prototype.pipe.call(this, dest, options);
};

module.exports = PrettyStream;
